### PR TITLE
Feature/reconstruct predict

### PR DESCRIPTION
**Pull request by ngorrinm**
This PR includes a significant refactor and enhancements to the `AutoEncoder` module.

**Model architecture and training**
- Added support for backward pass across RNN, LSTM, and GRU architectures.
- Integrated early stopping with customizable patience to prevent overfitting.
- Added optional post-decoder dense layer to allow smoother final compression and improved reconstruction output.
- Support for shuffling the training data, with configurable shuffle_buffer_size.
- Configurable batch_size to adjust training dynamics and memory usage.

**Flexible model configuration**
- Customizable number of hidden layers and dimensions.
- Optional bidirectionality for encoder and decoder.
- Support for variable activation functions per block.
- Feature selection for reconstruction (feature_to_check) with support for multiple values. 
- Issue #174 

**Data handling and preprocessing**
- Support for z-score normalization in addition to min-max scaling.
- Enhanced input flexibility by allowing both full datasets and explicit `(train, val, test)` tuple formats.
- Extended handling of datasets with missing values using masking strategies:
> - Automatic masks (set 0 for missing values and 1 for observed ones, derived directly from the data)
> - Custom masks (user-defined or based on NaN positions)
> - Added support for training directly on data with NaNs when masking is enabled, avoiding the need for prior imputation.

**Loss and evaluation**
- Refactored loss function to use masking, excluding missing or undesired values from loss calculation.
- Enabled the use of different weights per feature in the reconstruction error to reflect their relative importance.

**Reconstruct new data**
- Developed functionality to reconstruct new unseen data using a previously trained model.
- Iterative reconstruction process included to handle missing values through progressive imputation.
- Automatic plotting of actual vs reconstructed values, including per-ID visualizations.
- Generation and export of detailed reconstruction progress logs in case of nulls to Excel.

**Visualization and output**
- Improved plotting of original vs reconstructed series, now supporting multiple features, per-ID visualization, and custom feature labels.
- Added visual tracking of reconstruction progress over multiple iterations.

**Testing** 
- Reorganized the synthetic test module to support:
> - ID/no-ID scenarios
> - NaN values with different masking approaches
> - Tuple and non-tuple data inputs
> - Explicit control of masks per split (train/val/test)

### CODE REVIEW

File: mango_time_series/mango_time_series/models/autoencoder.py [hunk 1]
```diff
@@ -1,689 +1,1771 @@
...
+    def time_step_to_check(self) -> Optional[Union[int, List[int]]]:
+        """
+        Get the time step indices to check during reconstruction.
 
-        val_dataset = tf.data.Dataset.from_tensor_slices(self.x_val)
-        val_dataset = val_dataset.cache().batch(batch_size)
+        :return: Time step indices
+        :rtype: Optional[Union[int, List[int]]]
+        """
+        if not hasattr(self, "_time_step_to_check"):
```
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:165]: Why not initialize in the init like the other properties?

File: mango_time_series/mango_time_series/models/autoencoder.py [hunk 2]
```diff
@@ -1,689 +1,1771 @@
...
+        # If model is built, don't allow changes
+        if hasattr(self, "model") and self.model is not None:
+            raise ValueError(
+                "Cannot change time_step_to_check after model is built. "
+                "Call build_model() with the new time_step_to_check instead."
+            )
 
-        self.bidirectional_encoder = bidirectional_encoder
-        self.bidirectional_decoder = bidirectional_decoder
+        # Convert single integer to list
```
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:189]: I would enforce always to be a list then will be more consistent all over the code. Is there any reason?

Comment (antoniogonzalezsuarez) [reply mango_time_series/mango_time_series/models/autoencoder.py:189]: Maybe not for this PR but we could take a look into Pydantic to validate inputs in more complex codebases

File: mango_time_series/mango_time_series/models/autoencoder.py [hunk 3]
```diff
@@ -1,689 +1,1771 @@
...
-                    use_bidirectional=self.bidirectional_decoder,
-                    activation=self.activation_decoder,
-                    verbose=verbose,
-                ),
-            ],
-            name="autoencoder",
-        )
-        model.build()
+        # If context_window is set, validate indices are in range
+        if not hasattr(self, "_context_window"):
```
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:202]: Same as before with the init and just check for None here

File: mango_time_series/mango_time_series/models/autoencoder.py [hunk 4]
```diff
@@ -1,689 +1,1771 @@
...
+    def context_window(self) -> Optional[int]:
+        """
+        Get the context window size.
 
-        self.optimizer_name = optimizer
-        self.model_optimizer = self._get_optimizer(optimizer)
+        :return: Context window size or None if not initialized
+        :rtype: Optional[int]
+        """
+        return getattr(self, "_context_window", None)
```
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:221]: If we don't define at init we use this getattr to just return None, am i missing something?

I actually like to use ```__init__``` as kind of declaration of all attributes of the class

File: mango_time_series/mango_time_series/models/autoencoder.py [hunk 5]
```diff
@@ -1,689 +1,1771 @@
...
+        """
+        Set the activation function for the encoder.
+
+        :param value: Activation function
+        :type value: str
+        """
+        if hasattr(self, "model") and self.model is not None:
+            raise ValueError("Cannot change activation_encoder after model is built")
+
+        valid_activations = {
```
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:474]: Tryhard with a set for the optimization. LOVE IT :rocket:

File: mango_time_series/mango_time_series/models/autoencoder.py [hunk 6]
```diff
@@ -1,689 +1,1771 @@
...
+            if "model" not in saved_data or "params" not in saved_data:
+                raise ValueError("Invalid pickle file format: missing required keys.")
+
+            model = saved_data["model"]
+            params = saved_data["params"]
+
+            # Create an instance of AutoEncoder
+            instance = cls()
+
+            # Assign loaded parameters
```
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:1003]: Is there any reason to left ````__init__``` with no params and then set it like this? Is so we don´t have a long list of params in ```__init__```?

I guess is to make sure we run the validators in the setter but i think that they run if you do ```self.attribute = x``` in the ```__init__``` if setter is defined for attribute.

File: mango_time_series/mango_time_series/models/autoencoder.py [hunk 7]
```diff
@@ -1,689 +1,1771 @@
...
+            )
+
+        self.model = Sequential(self.layers, name="autoencoder")
+        self.model.build()
+        self.model_optimizer = optimizer
+
+        self._create_datasets(batch_size)
+
+    @property
+    def checkpoint(self) -> int:
```
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:1271]: Just style but this should be with the other properties

File: mango_time_series/mango_time_series/models/autoencoder.py [hunk 8]
```diff
@@ -1,689 +1,1771 @@
...
+        :type patience_counter: int
+        :param patience: Number of epochs to wait before early stopping
+        :type patience: int
+        :return: Tuple of (should_stop, best_val_loss, patience_counter)
+        :rtype: Tuple[bool, float, int]
+        """
+        if avg_val_loss < best_val_loss:
+            best_val_loss = avg_val_loss
+            patience_counter = 0
+            self.save(filename="best_model.pkl")
```
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:1556]: YOLO style I like it.

File: mango_time_series/mango_time_series/models/autoencoder.py [hunk 9]
```diff
@@ -692,81 +1774,1224 @@ def reconstruct(self):
...
+                norm_values = self.normalization_values["global"]
+
+                # Denormalize predictions
+                x_hat_train = denormalize_data(
+                    x_hat_train,
+                    self._normalization_method,
+                    min_x=norm_values["min_x"][self._feature_to_check],
+                    max_x=norm_values["max_x"][self._feature_to_check],
+                    mean_=(
+                        norm_values["mean_"][self._feature_to_check]
```
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:1801]: Not to be changed but you can use ```.get("mean_",{}).get('key',None)```

File: mango_time_series/mango_time_series/models/autoencoder.py
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/autoencoder.py:1]: In general, I have doubts of the use of ```__init__.py``` for this class and the "abuse" of ```hasattr``` and ```getattr```

File: mango_time_series/mango_time_series/models/utils/plots.py
Comment (antoniogonzalezsuarez) [inline mango_time_series/mango_time_series/models/utils/plots.py:1]: LGTM

**Review by antoniogonzalezsuarez — CHANGES_REQUESTED** [review]
Take a look at the comments

File: mango_time_series/mango_time_series/models/autoencoder.py (hunk 1 above)
Comment (gvalleg) [reply mango_time_series/mango_time_series/models/autoencoder.py:165]: We had doubts about the best way to do it. I will check and confirm with you.

File: mango_time_series/mango_time_series/models/autoencoder.py (hunk 3 above)
Comment (gvalleg) [reply mango_time_series/mango_time_series/models/autoencoder.py:202]: This is done because if you want to change the value of time_step_to_check it must meet the condition with respect to context_window and must have a value.

File: mango_time_series/mango_time_series/models/autoencoder.py (hunk 4 above)
Comment (gvalleg) [reply mango_time_series/mango_time_series/models/autoencoder.py:221]: If this is the best way to do it, we do it like this

File: mango_time_series/mango_time_series/models/autoencoder.py (hunk 1 above)
Comment (ngorrinm) [reply mango_time_series/mango_time_series/models/autoencoder.py:165]: Done

File: mango_time_series/requirements.txt
Comment (antoniogonzalezsuarez) [inline mango_time_series/requirements.txt:1]: If we merge my PR to develop we will have conflicts here

**Review by antoniogonzalezsuarez — APPROVED** [review]

**Review by gvalleg — APPROVED** [review]
